File: dailyQuestion/2020-05/05-27/python/solution_map.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


def detectCycle(head: ListNode) -> ListNode:
    tmp = {}
    while head:
        if head in tmp:
            return head
        tmp[head] = ""
        head = head.next
    return None


if __name__ == "__main__":
    node = ListNode(3)
    node.next = ListNode(2)
    node.next.next = ListNode(0)
    node.next.next.next = ListNode(-4)
    node.next.next.next.next = node.next
    result = detectCycle(node)
    print(result.val)

File: dailyQuestion/2020/2020-05/05-27/golang/solution_map.go
package main

import (
	"fmt"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func main() {
	node := &ListNode{Val: 3, Next: &ListNode{Val: 2}}
	node.Next.Next = &ListNode{Val: 0}
	node.Next.Next.Next = &ListNode{Val: -4}
	node.Next.Next.Next.Next = node.Next
	result := detectCycle(node)
	fmt.Println(result)

}

func detectCycle(head *ListNode) *ListNode {
	tmp := map[*ListNode]string{}
	for head != nil {
		if _, ok := tmp[head]; ok {
			return head
		}
		tmp[head] = ""
		head = head.Next
	}
	return nil
}


